target/linux/ipq40xx/image/Makefile

```diff
@@ -131,6 +131,22 @@ define Device/glinet_gl-b1300
 endef
 TARGET_DEVICES += glinet_gl-b1300
 
+define Device/linksys_ea6350v3
+	$(call Device/FitImage)
+	DEVICE_TITLE := Linksys EA6350v3
+	DEVICE_DTS := qcom-ipq4018-ea6350v3
+	BLOCKSIZE := 128k
+	PAGESIZE := 2048
+	KERNEL_SIZE := 3145728
+	IMAGE_SIZE := 38797312
+	UBINIZE_OPTS := -E 5
+	KERNEL = kernel-bin | gzip | fit gzip $$(DTS_DIR)/$$(DEVICE_DTS).dtb | append-uImage-fakehdr filesystem
```

Comment:
So the u-boot checking for a filesystem after the image.

If possible, could you please provide the u-boot's printenv? 
(Maybe there's a way around that)

Reply:
Here you have the fw_printenv output
`altkern=2800000
baudrate=115200
boot_count=0
boot_part_ready=3
boot_ver=0.6.0
bootcmd=if test $boot_part = 1; then run bootpart1; else run bootpart2; fi
bootdelay=2
bootpart1=set bootargs $partbootargs && nand read $loadaddr $prikern $kernsize && bootm $loadaddr
bootpart2=set bootargs $partbootargs2 && nand read $loadaddr $altkern $kernsize && bootm $loadaddr
ethact=eth0
ethaddr=60:38:E0:96:63:34
flash_type=0
flashimg=tftpboot $loadaddr $image && nand erase $prikern $imgsize && nand write $loadaddr $prikern $filesize
flashimg2=tftpboot $loadaddr $image && nand erase $altkern $imgsize && nand write $loadaddr $altkern $filesize
hw_mac_addr=60:38:E0:96:63:34
image=civic.img
imgsize=0x2800000
ipaddr=192.168.1.1
kernsize=300000
loadaddr=0x82000000
machid=8010100
netmask=255.255.255.0
partbootargs=init=/sbin/init rootfstype=squashfs ubi.mtd=11,2048 ubi.block=0,0 root=/dev/ubiblock0_0 rootwait ro
partbootargs2=init=/sbin/init rootfstype=squashfs ubi.mtd=13,2048 ubi.block=0,0 root=/dev/ubiblock0_0 rootwait ro
prikern=0
serverip=192.168.1.100
stderr=serial
stdin=serial
stdout=serial
sw_version=0.6.0
boot_part=2
auto_recovery=yes`
If you want the uboot's output... I cannot access the serial terminal nor the JTAG for the moment.

Reply:
WTF

```
kernsize=300000
loadaddr=0x82000000
prikern=0
altkern=2800000
bootpart1=set bootargs $partbootargs && nand read $loadaddr $prikern $kernsize && bootm $loadaddr
bootpart2=set bootargs $partbootargs2 && nand read $loadaddr $altkern $kernsize && bootm $loadaddr
bootcmd=if test $boot_part = 1; then run bootpart1; else run bootpart2; fi
```

This is really bad news.  The "nand read" command is reading from a fixed offset (either 0 or 2800000) with a fixed length (3MiB). This really just looks like that the uboot does not have support for UBI at all?! How is this even working then? Do they use some address->ubi wrapper like ASUS does or do they not use their bootcmd? What's going on?

Reply:
uboot does honor the bootcmd, as it's part of the "auto_recovery" feature and during a sysupgrade, the boot partition is effectively changed so I can confirm that bootcmd is being honored.

This likely explains this conversation at https://github.com/openwrt/openwrt/pull/1716/files/e9dc99ab2e1b243e64685ea85649e46d13874757#diff-39f92c7a5f7c48a3cf2183bd8b662a59

Reply:
> This really just looks like that the uboot does not have support for UBI at all?! How is this even working then? Do they use some address->ubi wrapper like ASUS does or do they not use their bootcmd? What's going on?
I've been fixing the DTS, and I found that the uboot does not support UBI. It boots a raw Linux Kernel from a fixed address. That means that this is not like the ASUS. In fact, the DTS was wrong. I fixed it like this:
```
	flash@1 {
		status = "okay";
		compatible = "spinand,mt29f", "w25n01gv";
		reg = <1>;
		spi-max-frequency = <24000000>;

		partitions {
			compatible = "fixed-partitions";
			#address-cells = <1>;
			#size-cells = <1>;

			kernel@0 {
				label = "kernel";
				reg = <0x00000000 0x02800000>;
			};
			rootfs@300000 {
				label = "rootfs";
				reg = <0x00300000 0x02800000>;
			};
			alt_kernel@2800000 {
				label = "alt_kernel";
				reg = <0x02800000 0x05000000>;
			};
			alt_rootfs@2b00000 {
				label = "alt_rootfs";
				reg = <0x02b00000 0x05000000>;
			};
			sysdiag@5000000 {
				label = "sysdiag";
				reg = <0x05000000 0x05100000>;
			};
			syscfg@5100000 {
				label = "syscfg";
				reg = <0x05100000 0x08000000>;
			};
			/* 0x00000000 - 0x08000000: 128 MiB */
		};
	};
```

Reply:
> > This really just looks like that the uboot does not have support for UBI at all?! How is this even working then? Do they use some address->ubi wrapper like ASUS does or do they not use their bootcmd? What's going on?
> 
> I've been fixing the DTS, and I found that the uboot does not support UBI. It boots a raw Linux Kernel from a fixed address. That means that this is not like the ASUS. In fact, the DTS was wrong. I fixed it like this:
> 
> ```
> 	flash@1 {
> 		status = "okay";
> 		compatible = "spinand,mt29f", "w25n01gv";
> 		reg = <1>;
> 		spi-max-frequency = <24000000>;
> 
> 		partitions {
> 			compatible = "fixed-partitions";
> 			#address-cells = <1>;
> 			#size-cells = <1>;
> 
> 			kernel@0 {
> 				label = "kernel";
> 				reg = <0x00000000 0x02800000>;

This might be a good time to look at: [Device Tree Usage: How Addressing Works](https://elinux.org/Device_Tree_Usage#How_Addressing_Works)

Do you see the #address-cells and #size-cells in the parent node?

> 			#address-cells = <1>;
> 			#size-cells = <1>;

This directly relate to the parameters of the "reg" properties in the child nodes:

So the first parameter in the "reg" is the address (or in your case the offset of the mtd partition) and it's one 32-bit value. The second parameter of the "reg" is the size (in your case the length of the partition).

So ```reg = <0x00000000 0x02800000>;``` means that the kernel partition starts at 0x0 and is 
41943040 = 40 MiB big! But from what I now know, I think you need ```reg = <0x00000000 0x00300000>;``` here because the kernel can only be 3 MiB big.

(Please note the kernel's dmesg MTD information uses start-offset and end-offset, so you can't directly copy the "end-offset" into the reg's size value. 

> 			};
> 			rootfs@300000 {
> 				label = "rootfs";
> 				reg = <0x00300000 0x02800000>;

maybe ```reg = <0x00300000 0x02500000>;``` => 37 MiB

> 			};

> 			alt_kernel@2800000 {
> 				label = "alt_kernel";
> 				reg = <0x02800000 0x05000000>;


maybe ```reg = <0x02800000 0x00300000>;```

> 			};
> 			alt_rootfs@2b00000 {
> 				label = "alt_rootfs";
> 				reg = <0x02b00000 0x05000000>;

maybe ```reg = <0x02b00000 0x02500000>;```

> 			};
> 			sysdiag@5000000 {
> 				label = "sysdiag";
> 				reg = <0x05000000 0x05100000>;
maybe ```reg = <0x05000000 0x00100000>;```
> 			};
> 			syscfg@5100000 {
> 				label = "syscfg";
> 				reg = <0x05100000 0x08000000>;
maybe ```reg = <0x05100000 0x2f00000>;```

> 			};
> 			/* 0x00000000 - 0x08000000: 128 MiB */
> 		};
> 	};
> ```

(I used "maybe" because I didn't really do the calculations with a piece of paper, so please double check the values!)

That said, The "stock" partition layout seems highly space-inefficient. Still I guess it will have to do.